File: ent/field/email/email_test.go
package email_test

import (
	"errors"
	"strings"
	"testing"

	"github.com/lolopinto/ent/ent/field/email"
	"github.com/stretchr/testify/require"
)

func TestDataType(t *testing.T) {
	testCases := map[string]struct {
		input  string
		output string
		err    error
	}{
		"simple":                     {"test@email.com", "test@email.com", nil},
		"trailing_space":             {"test@email.com ", "test@email.com", nil},
		"leading_space":              {" test@email.com", "test@email.com", nil},
		"leading_and_trailing_space": {" test@email.com ", "test@email.com", nil},
		"tabs":                       {"\ttest@email.com ", "test@email.com", nil},
		"with caps":                  {"Test@EMAIL.com ", "test@email.com", nil},
		"ALL CAPS":                   {"Test@EMAIL.com ", "test@email.com", nil},
		"with dash":                  {"first-last@EMAIL.com ", "first-last@email.com", nil},
		"with dots":                  {"first.last@EMAIL.com ", "first.last@email.com", nil},
		// bonus in the future, provide a way to distinugish these 2 for services which care about
		// preventing this usecase
		"with dots 2":     {"firstl.ast@EMAIL.com ", "firstl.ast@email.com", nil},
		"with gmail +":    {"test+spam-service@EMAIL.com ", "test+spam-service@email.com", nil},
		"with underscore": {"first_last@EMAIL.com", "first_last@email.com", nil},
		"no @":            {"foo.bar", "", errors.New("missing '@' or angle-addr")},
		// apparently, a valid email address
		"no.com": {"foo@bar", "foo@bar", nil},
		"name":   {"name <test@email.com>", "", errors.New("Invalid address")},
	}

	for key, tt := range testCases {
		t.Run(key, func(t *testing.T) {
			dt := email.Type()

			err := dt.Valid(tt.input)
			if tt.err == nil {
				require.Nil(t, err)
				res, err := dt.Format(tt.input)
				require.Nil(t, err)
				require.Equal(t, tt.output, res)
			} else {
				require.NotNil(t, err)
				require.Condition(t, func() bool {
					return strings.Contains(err.Error(), tt.err.Error())
				})
			}
		})
	}
}

File: ts/packages/ent-email/src/email.test.ts
import exp from "constants";
import { EmailType, EmailListType, Email } from "./email";

function testCase(exp: expectedResult) {
  let typ = EmailType({ name: "emailAddress" });
  if (exp.pre) {
    typ = exp.pre(typ);
  }
  if (exp.invalid) {
    expect(typ.valid(exp.input)).toBe(false);
    return;
  }

  expect(typ.valid(exp.input)).toBe(true);
  expect(typ.format(exp.input)).toBe(exp.output);
}

interface expectedResult {
  input: string;
  output?: string;
  invalid?: boolean;
  pre?: (typ: Email) => Email;
}

test("simple", () => {
  testCase({
    input: "test@email.com",
    output: "test@email.com",
  });
});

test("trailing space", () => {
  testCase({
    input: "test@email.com ",
    output: "test@email.com",
  });
});

test("leading space", () => {
  testCase({
    input: " test@email.com",
    output: "test@email.com",
  });
});

test("tabs", () => {
  testCase({
    input: "\ttest@email.com",
    output: "test@email.com",
  });
});

test("with capps", () => {
  testCase({
    input: "Test@Email.com",
    output: "test@email.com",
  });
});

test("all caps", () => {
  testCase({
    input: "TEST@EMAIL.com",
    output: "test@email.com",
  });
});

test("with dash", () => {
  testCase({
    input: "first-last@email.com",
    output: "first-last@email.com",
  });
});

test("with email", () => {
  testCase({
    input: "first.last@email.com",
    output: "first.last@email.com",
  });
});

test("with missing @", () => {
  testCase({
    input: "test.email.com",
    invalid: true,
  });
});

test("no .com", () => {
  testCase({
    input: "first.last@email",
    output: "first.last@email",
  });
});

test("with gmail+", () => {
  testCase({
    input: "test+spam@email.com",
    output: "test+spam@email.com",
  });
});

test("with underscore", () => {
  testCase({
    input: "first_last@email.com",
    output: "first_last@email.com",
  });
});

test("with name", () => {
  testCase({
    input: "first last <first.last@email.com>",
    invalid: true,
  });
});

test("with comments", () => {
  testCase({
    input: "first last (comment) <first.last@email.com>",
    invalid: true,
  });
});

test("restrict domain", () => {
  testCase({
    pre: (typ) => typ.domain("email.com"),
    input: "first_last@email.com",
    output: "first_last@email.com",
  });
});

test("restrict domain. invalid ", () => {
  testCase({
    pre: (typ) => typ.domain("email.com"),
    input: "first_last@bar.com",
    invalid: true,
  });
});

test("list", () => {
  const tt = EmailListType({ name: "emails" });
  const input = ["first_last@email.com", "first_last@bar.com"];
  expect(tt.valid(input)).toBe(true);
  // postgres stored in db style
  expect(tt.format(input)).toEqual(`{${input.join(",")}}`);
});

test("list with invalid", () => {
  const tt = EmailListType({ name: "emails" });
  const input = [
    "first_last@email.com",
    "first_last@bar.com",
    "  first last(comment) < first.last@email.com>",
  ];
  expect(tt.valid(input)).toBe(false);
});
